### PR TITLE
AWS deployment script

### LINKED ISSUE #706: Suggest a Revision in Makefile Section amazon-cluster
A suggestion is given in this issue to revise the `Makefile` in order to be compatible with AWS client.

The original script would lead to a fatal error `IllegalLocationConstraintException`

```
amazon-cluster:
        #aws s3api create-bucket --bucket $(BUCKET_NAME) --region $(AWS_REGION)
        aws ec2 create-volume --size $(STORAGE_SIZE) --region $(AWS_REGION) --availability-zone $(AWS_AVAILABILITY_ZONE) --volume-type gp2
```

I suggest to revise the script to

```
amazon-cluster:
        aws s3api create-bucket --bucket $(BUCKET_NAME) --create-bucket-configuration LocationConstraint=$(AWS_REGION)
        aws ec2 create-volume --size $(STORAGE_SIZE) --region $(AWS_REGION) --availability-zone $(AWS_AVAILABILITY_ZONE) --volume-type gp2
```

The new script has been checked and it works for following process.
Similar issues are also reported to aws_cli team. The reference here (https://github.com/aws/aws-cli/issues/589)
The option used in the original script is not allowed in aws client reference. http://kubernetes.io/docs/user-guide/kubectl/kubectl_create/

### LINKED ISSUE #708: Suggest a Revision in Makefile Section clean-amazon-cluster
The `Makefile`  is not effective enough in the section `clean-amazon-cluster`. Several components remained undeleted after `make clean-launch` and `make clean-amazon-cluster`. I suggest to add a `clean-amazon-launch` section with the following script

```
clean-launch
clean-amazon-cluster
kubectl delete secrets amazon-secret
kubectl delete persistentvolumes rethink-volume
kubectl delete persistentvolumeclaims rethink-volume-claim
```

The script has been merged in my program locally and it works to deploy amazon.

### PR DESCRIPTION
**Pull request by ShengjieLuo**
A request is pulled to revise the `Makefile` in order to be compatible with AWS client.

First Part:
The original script would lead to a fatal error `IllegalLocationConstraintException`
```
amazon-cluster:
        #aws s3api create-bucket --bucket $(BUCKET_NAME) --region $(AWS_REGION)
        aws ec2 create-volume --size $(STORAGE_SIZE) --region $(AWS_REGION) --availability-zone $(AWS_AVAILABILITY_ZONE) --volume-type gp2
```

I suggest to revise the script to
```
amazon-cluster:
        aws s3api create-bucket --bucket $(BUCKET_NAME) --create-bucket-configuration LocationConstraint=$(AWS_REGION)
        aws ec2 create-volume --size $(STORAGE_SIZE) --region $(AWS_REGION) --availability-zone $(AWS_AVAILABILITY_ZONE) --volume-type gp2
```

The new script has been checked and it works for following process.
Similar issues are also reported to aws_cli team. The reference here (https://github.com/aws/aws-cli/issues/589)
The option used in the original script is not allowed in aws client reference. http://kubernetes.io/docs/user-guide/kubectl/kubectl_create/

fix #706 

Second Part:
The Makefile is not effective enough in the section clean-amazon-cluster. Several components remained undeleted after make clean-launch and make clean-amazon-cluster. I suggest to add a amazon-clean-launch section to clearly delete the pachyderm component. Also the readme.md is revised to display the change.

fix #708 

Note:  Some version mismatching occurred in my commits, it has been corrected in the final one. Thanks.

### CODE REVIEW

File: doc/deploying_setup.md [hunk 1]
```diff
@@ -184,7 +184,13 @@ $ AWS_ID=[access key ID] AWS_KEY=[secret access key] AWS_TOKEN=[session token] m
 $ make MANIFEST=manifest launch
 ```
 
-It may take a while to complete for the first time, as a lot of Docker images need to be pulled.
+It may take a while to complete for the first time, as a lot of Docker images need to be pulled. User can tear down the pachyderm bu running:
```
Comment (zyluo) [inline doc/deploying_setup.md]: bu -> by

File: doc/deploying_setup.md [hunk 2]
```diff
@@ -184,7 +184,13 @@ $ AWS_ID=[access key ID] AWS_KEY=[secret access key] AWS_TOKEN=[session token] m
...
+
+```shell
+$ make amazon-clean$
+```
+
+It may take about 30 seconds to delete the pachyderm components.
```
Comment (zyluo) [inline doc/deploying_setup.md]: Pachyderm

File: Makefile [hunk 3]
```diff
@@ -259,13 +259,22 @@ amazon-cluster-manifest:
 	@pach-deploy amazon $(BUCKET_NAME) $(AWS_ID) $(AWS_KEY) $(AWS_TOKEN) $(AWS_REGION) $(STORAGE_NAME) $(STORAGE_SIZE)
 
 amazon-cluster:
-	aws s3api create-bucket --bucket $(BUCKET_NAME) --region $(AWS_REGION)
+	aws s3api create-bucket --bucket $(BUCKET_NAME) --create-bucket-configuration LocationConstraint=$(AWS_REGION)
```
Comment (jdoliner) [inline Makefile]: In the example from `aws s3ap create-bucket help` they have both `LocationConstraint` and `--region` set. Does it work with only the `LocationConstrain`?

Comment (ShengjieLuo) [inline Makefile]: See the amazon reference here http://docs.aws.amazon.com/cli/latest/reference/s3api/create-bucket.html
The `--create-bucket-configuration LocationConstraint=$(AWS_REGION)` is recommended by aws if we use the region besides the default region `us-west-1`. 
image

Also the guide doesn't give an option named '--region <regionName>'., even sometimes `--region <regionName>` still can be used.
image

In conclusion, I recommend to adapt the script to `--create-bucket-configuration LocationConstraint=$(AWS_REGION)`

File: doc/deploying_setup.md (hunk 1 above)
Comment (ShengjieLuo) [inline doc/deploying_setup.md]: I would correct the typo error soon. Thanks for the notice.

**Comment by jdoliner** [comment]
@ShengjieLuo that's the documentation I did read. As you can see all of their commands have `--region` even if they also use `LocationConstraint`. That's what I'm worried about.

File: Makefile (hunk 3 above)
Comment (ShengjieLuo) [inline Makefile]: @jdoliner  For your worry, I do some additional tests to justify my request. I used a new aws account to do the following experiments with the default-region us-west-2.
The result is here, and a screenshot is shown to describe the situation.
```
# aws s3api create-bucket --bucket shengjieTest1 --region us-west-2
An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation:
The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.
# aws s3api create-bucket --bucket shengjieTest1 --region us-east-1
{
    "Location": "/shengjieTest1"
}
# aws s3api create-bucket --bucket shengjietest2 --create-bucket-configuration LocationConstraint=us-west-2
{
    "Location": "http://shengjietest2.s3.amazonaws.com/"
}
# aws s3api create-bucket --bucket shengjietest5 --region us-west-2
{
    "Location": "/shengjietest5"
}
# aws s3api create-bucket --bucket shengjieTest4 --create-bucket-configuration LocationConstraint=us-west-2
An error occurred (InvalidBucketName) when calling the CreateBucket operation: The specified bucket is not valid.
#aws s3api create-bucket --bucket shengjieTest4 --create-bucket-configuration LocationConstraint=us-east-1
An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The us-east-1 location constraint is incompatible for the region specific endpoint this request was sent to.
```
In conclusion, we summarize the error response here

1.  The s3-bucket name is allowed to involve the uppercase letter only if the location of s3-bucket is `us-east-1` with the `create-bucket region --bucket <name> --region <region_name> `
2.  If we establish the s3-bucket with a name involving the uppercase lestter

- 2.1 within the `--create-bucket-configuration <default-region>`, we have the error response as `InvalidBucketName` 
- 2.2 within the `--create-bucket-configuration <none-default-region>` we have the error response as `IllegalLocationConstraintException`
- 2.3 within the `--region <region_other _than_us-east-1>` we have the error response as `IllegalLocationConstraintException`
- 2.4 within the `--region <us-east-1>` , no error response, the bucket can be established.

If the bucket name follows the following points, than it can be used in both options.

- Bucket names should not contain upper-case letters
- Bucket names should not contain underscores (_)
- Bucket names should not end with a dash
- Bucket names should be between 3 and 63 characters long
- Bucket names cannot contain dashes next to periods (e.g., my-.bucket.com and my.-bucket are invalid)
- Bucket names cannot contain periods

Therefore, from my opinion, a troubleshooting issue is better than revising the code. And the original code don't need to be revised. Please focus on the second part `make amazon-clean`. Thanks.
